dts/arm/ambiq/ambiq_apollo4p.dtsi

```diff
@@ -104,6 +105,44 @@
 		pinctrl: pin-controller@40010000 {
 			compatible = "ambiq,apollo4-pinctrl";
 			reg = <0x40010000 0x800>;
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			gpioa: gpioa@0 {
```

Comment:
After looking at the HAL (sigh), and its SVD files, I see that hw engineers had fun by offsetting per-register instead of creating multiple independent ports.

```
    <peripheral>
      <name>FPIO</name>
      <version>1.0</version>
      <description>Fast PIO access</description>
      <!-- <groupName>GROUP_FPIO</groupName> -->
      <baseAddress>0x48001000</baseAddress>
      <size>32</size>
      <access>read-write</access>

      <addressBlock>
        <offset>0</offset>
        <size>0x00000070</size>
        <usage>registers</usage>
      </addressBlock>

      <registers>
        <register>
          <name>RD0</name>
          <description>GPIO Input 0 (31-0)</description>
          <addressOffset>0x00000000</addressOffset>
          <size>32</size>
          <access>read-write</access>
          <resetValue>0x00000000</resetValue>
          <resetMask>0xFFFFFFFF</resetMask>

          <fields>
            <field>
              <name>RD0</name>
              <description>GPIO31-0 Reads pin state - read only. Returns the pad pin state for pins 0-31 if the PINCFG's input enable (INPEN) is active and RDZERO is inactive.</description>
              <bitRange>[31:0]</bitRange>
              <access>read-write</access>

            </field>
          </fields>
        </register>
        <register>
          <name>RD1</name>
          <description>GPIO Input 1 (63-32)</description>
          <addressOffset>0x00000004</addressOffset>
          <size>32</size>
          <access>read-write</access>
          <resetValue>0x00000000</resetValue>
          <resetMask>0xFFFFFFFF</resetMask>

          <fields>
            <field>
              <name>RD1</name>
              <description>GPIO63-32 Reads pin state - read only. Returns the pad pin state for pins 0-31 if the PINCFG's input enable (INPEN) is active and RDZERO is inactive.</description>
              <bitRange>[31:0]</bitRange>
              <access>read-write</access>

            </field>
          </fields>
        </register>
        <register>
          <name>RD2</name>
          <description>GPIO Input 2 (95-64)</description>
          <addressOffset>0x00000008</addressOffset>
          <size>32</size>
          <access>read-write</access>
          <resetValue>0x00000000</resetValue>
          <resetMask>0xFFFFFFFF</resetMask>

          <fields>
            <field>
              <name>RD2</name>
              <description>GPIO95-64 Reads pin state - read only. Returns the pad pin state for pins 0-31 if the PINCFG's input enable (INPEN) is active and RDZERO is inactive.</description>
              <bitRange>[31:0]</bitRange>
              <access>read-write</access>

            </field>
          </fields>
        </register>
        <register>
          <name>RD3</name>
          <description>GPIO Input 3 (127-96)</description>
          <addressOffset>0x0000000C</addressOffset>
          <size>32</size>
          <access>read-write</access>
          <resetValue>0x00000000</resetValue>
          <resetMask>0xFFFFFFFF</resetMask>

          <fields>
            <field>
              <name>RD3</name>
              <description>GPIO127-96 Reads pin state - read only. Returns the pad pin state for pins 0-31 if the PINCFG's input enable (INPEN) is active and RDZERO is inactive.</description>
              <bitRange>[31:0]</bitRange>
              <access>read-write</access>

            </field>
          </fields>
        </register>
        <register>
          <name>WT0</name>
          <description>GPIO Output 0 (31-0)</description>
          <addressOffset>0x00000010</addressOffset>
          <size>32</size>
          <access>read-write</access>
          <resetValue>0x00000000</resetValue>
          <resetMask>0xFFFFFFFF</resetMask>

          <fields>
            <field>
              <name>WT0</name>
              <description>GPIO31-0 Reads or writes pin state. Writes of 1 bits set output pad signal if the GPIO is enabled for output. Reads return status, including sets/clears through the WTS and WTC registers.</description>
              <bitRange>[31:0]</bitRange>
              <access>read-write</access>

            </field>
          </fields>
        </register>
```

and in the HAL you can clearly see it, e.g.

```
#define AM_HAL_GPIO_RDn(pin)    ((volatile uint32_t *)&GPIO->RD0  + (((pin) >> 5) & 0x3))
```

So technically speaking, one has 1 port and 32*4 pins. I wonder then why we have `gpioa`, `gpiob`, etc in the devicetree. I guess Ambiq users will find this weird, I can't see any gpioa/b/etc in the sample code they have.

Reply:
I don't really understand what you mean by that. As I said:
>  The answer to why we use this pin offset is that the Ambiq GPIO is actually a single instance with 128 pins.

The whole discussion was about this limitation. Can you please let me know if the current workaround the API limitation is OK with you, and if not - how should I approach it differently?

Reply:
@msierszulski Ambiq has done testing with the code from this pull request.
It works with the apollo4p_evb and looks to be inline with the rest of the Zephyr ecosystem from our point of view.
Thanks for the good work!